File: odemapster-core/src/main/java/es/upm/fi/dia/oeg/obdi/core/materializer/NTripleMaterializer.java
package es.upm.fi.dia.oeg.obdi.core.materializer;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;

import org.apache.log4j.Logger;

import com.hp.hpl.jena.rdf.model.AnonId;
import com.hp.hpl.jena.rdf.model.Model;
import com.hp.hpl.jena.rdf.model.RDFNode;
import com.hp.hpl.jena.vocabulary.RDF;

import es.upm.fi.dia.oeg.morph.base.GeneralUtility;


public class NTripleMaterializer extends AbstractMaterializer {
	private String currentSubject;
	private static Logger logger = Logger.getLogger(NTripleMaterializer.class);
	private Writer writer;

	public NTripleMaterializer(String outputFileName, Model model) throws IOException {
		this.outputFileName = outputFileName;
		this.model = model;
		//this.writer = new OutputStreamWriter (new FileOutputStream (outputFileName), "UTF-8");
	}

	private void write(String triple) throws Exception {
		if(this.writer == null) {
			this.writer = new OutputStreamWriter (new FileOutputStream (outputFileName), "UTF-8");
		}
		writer.append(triple);
	}

	@Override
	public
	void materializeRDFTypeTriple(String subjectURI, String conceptName,
			boolean isBlankNodeSubject, String graph) {
		String triple = null;
		try {
			this.currentSubject = this.createSubject(isBlankNodeSubject, subjectURI);
			//this.currentSubject = this.currentSubject.replaceAll("\n","").replaceAll("\r", "");

			triple = 
					GeneralUtility.createQuad(
							this.currentSubject
							, GeneralUtility.createURIref(RDF.type.toString())
							, GeneralUtility.createURIref(conceptName)
							, GeneralUtility.createURIref(graph)
							); 

			//writer.append(triple);
			this.write(triple);
		} catch(Exception e) {
			logger.error("unable to serialize triple, subjectURI=" + subjectURI);
		}

	}

	@Override
	public void materializeDataPropertyTriple(String predicateName,
			Object propVal, String datatype,
			String lang, String graph) {
		String triplePredicate = GeneralUtility.createURIref(predicateName);

		String propValString = propVal.toString();

		String literalString = null;
		if(datatype == null) {
			if(lang == null) {
				literalString = GeneralUtility.createLiteral(propValString);
			} else {
				literalString = GeneralUtility.createLanguageLiteral(propValString, lang);
			}
		} else {
			literalString = GeneralUtility.createDataTypeLiteral(propValString, datatype);
		}

		String tripleString = null; 
		try {
			if(this.currentSubject != null) {
				tripleString = GeneralUtility.createQuad(this.currentSubject, triplePredicate, literalString, GeneralUtility.createURIref(graph));
				//writer.append(tripleString);
				this.write(tripleString);
			}
		} catch(Exception e) {
			logger.error("unable to serialize triple : " + tripleString + " because " + e.getMessage());
		}
	}



	@Override
	public void materializeObjectPropertyTriple(String predicateName,
			String rangeURI, boolean isBlankNodeObject, String graph) {

		String objectString;
		if(isBlankNodeObject) {
			objectString = GeneralUtility.createBlankNode(rangeURI);
		} else {
			objectString = GeneralUtility.createURIref(rangeURI);
		}

		String triple = null;
		try {
			if(this.currentSubject != null) {
				triple = GeneralUtility.createQuad( this.currentSubject
						, GeneralUtility.createURIref(predicateName)
						, objectString, GeneralUtility.createURIref(graph)); 
				this.write(triple);				
			}
		} catch(Exception e) {
			logger.error("unable to serialize triple : " + triple);
		}

	}


	@Override
	public String createSubject(boolean isBlankNode, String subjectURI) {
		if(isBlankNode) {
			AnonId anonId = new AnonId(subjectURI);
			model.createResource(anonId);		
		} else {
			model.createResource(subjectURI);
		}

		if(isBlankNode) {
			this.currentSubject = GeneralUtility.createBlankNode(subjectURI);
		} else {
			this.currentSubject = GeneralUtility.createURIref(subjectURI);
			
			try {
				GeneralUtility.substituteEntitiesInElementContent(this.currentSubject);
			} catch(Exception e) {
				logger.warn("Not well formed address : " + this.currentSubject);
			}
		}
		return this.currentSubject;
	}


	@Override
	public void materialize() throws IOException {
		//nothing to do, the triples were added during the data translation process
		if(this.writer != null) {
			//this.writer.flush();
			this.writer.close();
		}
	}
	
//	@Override
//	public void materializeQuad(String subject, String predicate, String object, String graph) {
//		String quad = GeneralUtility.createQuad(subject, predicate, object, graph);
//		try {
//			this.write(quad);
//		} catch(Exception e) {
//			logger.error("unable to serialize triple, subjectURI=" + quad);
//		}
//	}

	@Override
	public void materializeQuad(RDFNode subject, RDFNode predicate,
			RDFNode object, RDFNode graph) {
		String triple = null;
		try {
			String subjectString = GeneralUtility.nodeToString(subject);
			String predicateString = GeneralUtility.nodeToString(predicate);
			String objectString = GeneralUtility.nodeToString(object);
			String graphString = GeneralUtility.nodeToString(graph);
			
			triple = GeneralUtility.createQuad(subjectString, predicateString, objectString, graphString);
			this.write(triple);
		} catch(Exception e) {
			logger.error("unable to serialize triple, subjectURI=" + subject);
		}
		
	}	
}

File: morph-base/src/main/scala/es/upm/fi/dia/oeg/morph/base/materializer/NTripleMaterializer.scala
package es.upm.fi.dia.oeg.morph.base.materializer

//import com.hp.hpl.jena.rdf.model.Model
import org.apache.jena.graph.Node
import org.apache.jena.rdf.model.Model
import org.apache.jena.rdf.model.impl.PropertyImpl
import org.apache.jena.riot.Lang
import org.apache.jena.sparql.core.Quad;
//import org.apache.log4j.Logger
import java.io.Writer
import java.io.OutputStreamWriter
import java.io.FileOutputStream
//import com.hp.hpl.jena.rdf.model.RDFNode
import org.apache.jena.rdf.model.RDFNode;
import es.upm.fi.dia.oeg.morph.base.GeneralUtility
import java.io.OutputStream
import java.io.PrintWriter
import java.io.BufferedOutputStream
import java.io.BufferedWriter
//import com.hp.hpl.jena.rdf.model.Property
import org.apache.jena.rdf.model.Property;
import org.slf4j.LoggerFactory

import org.apache.jena.riot.system.StreamOps
import org.apache.jena.riot.system.StreamRDF
import org.apache.jena.riot.system.StreamRDFWriter
import scala.collection.JavaConversions._
import scala.collection.JavaConverters._

class NTripleMaterializer(model:Model,ntWriter:Writer, ntOutputStream: OutputStream)
  extends MorphBaseMaterializer(model,ntWriter, ntOutputStream) {
  //THIS IS IMPORTANT, SCALA PASSES PARAMETER BY VALUE!
  this.writer = ntWriter;
  this.outputStream = ntOutputStream


  override 	val logger = LoggerFactory.getLogger(this.getClass());


  def write(triple:String ) = {
    this.writer.write(triple)
    this.writer.flush();
  }

  def streamQuad(quad:Quad) = {
    val graphNode = quad.getGraph;
    if(graphNode == null) {
      val triple = quad.asTriple()
      val streamRDF:StreamRDF = StreamRDFWriter.getWriterStream(outputStream, Lang.NTRIPLES)
      streamRDF.start()
      val tripleIterator = List(triple).toIterator.asJava;
      StreamOps.sendTriplesToStream(tripleIterator, streamRDF)
      streamRDF.finish()
    } else {
      val quadIterator = List(quad).toIterator.asJava;
      val streamRDF:StreamRDF = StreamRDFWriter.getWriterStream(outputStream, Lang.NQUADS)
      streamRDF.start()
      StreamOps.sendQuadsToStream(quadIterator, streamRDF)
      streamRDF.finish()
    }

  }

  override def materialize() {
    //nothing to do, the triples were added during the data translation process
    if(this.writer != null) { this.writer.flush(); }
    if(this.outputStream != null) {
      this.outputStream.flush()
      this.outputStream.close()
    }

  }

  override def materializeQuad(subjectNode:Node, predicateNode:Node, objectNode:Node, graphNode:Node) {
    if(subjectNode != null && predicateNode != null && objectNode != null) {
      try {
        /*
				val subjectString = GeneralUtility.nodeToString(subject);
				val predicateString = GeneralUtility.nodeToString(predicate);
				val objectString = GeneralUtility.nodeToString(obj);
				val graphString = GeneralUtility.nodeToString(graph);
				val tripleString = GeneralUtility.createQuad(subjectString, predicateString, objectString, graphString);
				this.write(tripleString);
				*/

        /*
        val graphNode = if(graph != null) { graph.asNode() } else { null }
        val subjectNode = if(subject != null) { subject.asNode() } else { null }
        val objectNode = if(obj != null) { obj.asNode() } else { null }
        */
        //val predicateNode = if(predicate != null) { predicate.asNode() } else { null }
        val quad = Quad.create(graphNode, subjectNode, predicateNode, objectNode)
        this.streamQuad(quad)

      } catch {
        case e:Exception => {
          //e.printStackTrace();
          val errorMessage = "unable to serialize triple, subjectURI=" + subjectNode + ", error message = " + e.getMessage();
          logger.debug(errorMessage);
          //noOfErrors = noOfErrors + 1;
        }
      }
    } else {
      if(subjectNode == null) {
        val errorMessage = "unable to serialize triple, subject is null!";
        logger.debug(errorMessage);
      }

      if(predicateNode == null) {
        val errorMessage = "unable to serialize triple, predicate is null!";
        logger.debug(errorMessage);
      }

      if(objectNode == null) {
        val errorMessage = "unable to serialize triple, object is null!";
        //logger.debug(errorMessage);
      }


    }
  }

  //	override def postMaterialize() = {
  ////		if(this.writer != null) {
  ////			this.writer.flush();
  ////			this.writer.close();
  ////		}
  //	}
}